### PR TITLE
F/libmongo client to mongo c driver

### LINKED ISSUE #361: mongodb: moving to official C mongo client library
https://github.com/mongodb/mongo-c-driver

(as algernon said, problems that leads to creating an alternative libmongo-client are not exist anymore)

### PR DESCRIPTION
**Pull request by u2yg**
Fixes #361

### CODE REVIEW

**Comment by bazsi** [comment]
It seems ok, however this is too complex for me to review especially as I don't really know mongodb.

**Comment by u2yg** [comment]
@ihrwein I have supposedly pushed fixes to almost all of your questions. The last trivial change needed is whether we should remove the redundant `self->recovery_cache = NULL` or not.

File: .gitmodules [hunk 1]
```diff
@@ -10,3 +7,7 @@
 [submodule "modules/afamqp/rabbitmq-c"]
 	path = modules/afamqp/rabbitmq-c
 	url = https://github.com/alanxz/rabbitmq-c.git
+[submodule "modules/afmongodb/mongo-c-driver"]
```
Comment (lbudai) [inline .gitmodules]: Do we really need `mongo-c-driver` to be included as a submodule?

File: modules/afmongodb/afmongodb.c [hunk 2]
```diff
@@ -611,9 +711,9 @@ afmongodb_dd_init(LogPipe *s)
           for (l=self->servers; l; l = g_list_next(l))
             {
               gchar *host = NULL;
-              gint port = 27017;
+              gint port = MONGOC_DEFAULT_PORT;
```
Comment (lbudai) [inline modules/afmongodb/afmongodb.c]: Based on this, `MONGOC_DEFAULT_PORT` cannot be used in place of `MONGO_CONN_LOCAL`.
MongoDB provides a way for local connections via a unix domain socket. When  `MONGO_CONN_LOCAL` was set as port, then we connect to mongodb via unix domain socket (value for `MONGO_CONN_LOCAL` is set for -1 in the libmongo client) .

**Comment by u2yg** [comment]
We've now decided that major reworking will be done on this implementation (connection string in config file). Please no more reviews until then.

File: .gitmodules (hunk 1 above)
Comment (u2yg) [inline .gitmodules]: I'm not sure. Fortunately for us, they have just released the 1.2 stable few days ago. Unfortunately, I found and fixed a UNIX domain socket bug in mongo-c-driver and I'm not sure when they will merge my pull request. I don't think this is an import issue, we could simply disregard that for now and use something stable.

**Comment by u2yg** [comment]
@lbudai, @ihrwein: You can now review the last commit which supersedes and simplifies many parts of the previous implementation. It's not urgent, as we've agreed in that it is only going to be merged into 3.8. I just wasn't sure whether GitHub sends out mails about label adjustments.

File: .gitmodules (hunk 1 above)
Comment (bkil-syslogng) [inline .gitmodules]: The domain socket fix has been merged under tag 1.2.1 along with the other one:
https://github.com/mongodb/mongo-c-driver/commit/6614693e481b3614a8ae8af2dc10bac2c768a82a

**Comment by lbudai** [comment]
@bkil-syslogng : it is ready for merge, but before merging this, I have to prepare my OBS environment, etc... (we will get back to this in January :-) )

**Comment by bkil-syslogng** [comment]
If you are fine with the implementation after reviewing, I shall:
* update the mongo-c-driver version to the most recent stable tag;
* retest;
* rebase to master;
* rebase/squash commits to a much fewer number

**Comment by lbudai** [comment]
@bkil-syslogng : could you rebase it against master?
Then I'd merge it.

**Comment by bkil-syslogng** [comment]
Rebased and resubmitted under https://github.com/balabit/syslog-ng/issues/891

**Comment by lbudai** [comment]
thanks

**Pull request closed without merging**